File: lab02/lab02.py
#!/usr/bin/env python
# coding: utf-8

# # Time to solve another problem together! Here's the problem statement:
# 
# Given two strings s and t, return true if t is an anagram of s, and false otherwise.
# 
# An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

# In[36]:


def createHashMap(input):
    hashmap = {}
    for x in input:
        hashmap[x] = hashmap.get(x, 0) + 1
    return hashmap

def anagram(s,t):
    return False if len(s) != len(t) or  createHashMap(s) != createHashMap(t) else True


# In[37]:


s,t = "doga" , "goad"
anagram(s,t)


# In[38]:


test_cases = [
    ("listen", "silent"),
    ("hello", "hello"),
    ("earth", "hearted"),
    ("Tea", "eAt"),
    ("debit card", "bad credit"),
    ("astronomer", "moon starer"),
    ("Café", "éaFc"),
    ("", ""),
    ("a", "a"),
    ("hello", "world")
]

GREEN = '\033[92m'
RED = '\033[91m'
RESET = '\033[0m'

for idx, (str1, str2) in enumerate(test_cases, start=1):
    result = 'True' if anagram(str1, str2) else 'False'
    color = GREEN if result == 'True' else RED
    print(f"Test Case {idx}: {color}{result}{RESET} : '{str1}' and '{str2}' ")


# In[39]:


get_ipython().system('jupyter nbconvert --to script lab02.ipynb')

File: lab02/lab02.java
package lab02;

import java.util.HashMap;
import java.util.Map;

public class lab02 {
    
    public static Map<Character, Integer> createHashMap(String input) {
        Map<Character, Integer> hashmap = new HashMap<>();
        for (char x : input.toCharArray()) {
            hashmap.put(x, hashmap.getOrDefault(x, 0) + 1);
        }
        return hashmap;
    }

    public static boolean anagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }
        Map<Character, Integer> sMap = createHashMap(s);
        Map<Character, Integer> tMap = createHashMap(t);
        return sMap.equals(tMap);
    }

    public static void main(String[] args) {
        String[][] testCases = {
            {"listen", "silent"},
            {"hello", "hello"},
            {"earth", "hearted"},
            {"Tea", "eAt"},
            {"debit card", "bad credit"},
            {"astronomer", "moon starer"},
            {"Café", "éaFc"},
            {"", ""},
            {"a", "a"},
            {"hello", "world"}
        };

        String GREEN = "\u001B[92m";
        String RED = "\u001B[91m";
        String RESET = "\u001B[0m";

        for (int idx = 0; idx < testCases.length; idx++) {
            String str1 = testCases[idx][0];
            String str2 = testCases[idx][1];
            boolean result = anagram(str1, str2);
            String color = result ? GREEN : RED;
            System.out.printf("Test Case %d: %s%s%s : '%s' and '%s'\n", idx + 1, color, result, RESET, str1, str2);
        }
    }
}
